### PR TITLE
Update to 0.7

### PR DESCRIPTION
**Pull request by pablosanjose**
Fix for #59, alternative approach to #60 

Since `ReinterpretArray` seems to have an unacceptable indexing overhead here (see #60 and https://discourse.julialang.org/t/big-overhead-with-the-new-lazy-reshape-reinterpret/7635/1), @KristofferC advices to `copy` instead of `reshape(reinterpret(data...),...)`. 

This PR implements two parts required for this approach:

1- A `@noinline reinterpret_or_copy` function that uses `reinterpret` for julia versions prior to the introduction of `ReinterpretArray` lazy wrappers, and makes a reinterpreted copy otherwise.

2- Change from the deprecated uninitialized contructor `Array{T}(::Int...)` to the new form `Array{T}(uninitialized, ::Int...)`. Not doing this introduced a very large performance regression under julia master. To remain backward compatible I used `@compat`, which adds a dependency to Compat.jl.

With this, the timings for a few tests I tried were identical to the ones in the master branch under v0.6.

### CODE REVIEW

File: test/scratch.jl [hunk 1]
```diff
@@ -0,0 +1,189 @@
+NearestNeighbors.knn_point!(tree, [.1,.2,.3], false, Vector{Float64}(5), Vector{Int}(5), NearestNeighbors.always_false)
```
Comment (KristofferC) [inline test/scratch.jl]: Accidentally committed this file?

Comment (pablosanjose) [reply test/scratch.jl]: Oops, yes! Removing...

**Comment by pablosanjose** [comment]
A question: should I revert the `uninitialized` commit and make a separate PR? They are quite orthogonal fixes, although both are required to restore the performance we had before.

**Comment by KristofferC** [comment]
Nah, just keep it in the same :)

**Comment by KristofferC** [comment]
Currently blocked on StaticArrays 0.7 support.

**Comment by KristofferC** [comment]
Updated this.

I couldn't get the `reorder` thing to work with the data free trees so, for now, I removed it. @rened Is this feature important to you?

**Comment by rened** [comment]
No worries, fine with me!

**Pull request merged**